Completion tracking for accurate upload and download progress reporting.

hoytak opened the pull request:
This PR adds detailed completion tracking to the upload and download paths that reports accurate per-file metrics to the progress updating interface.  The dependencies between xorbs and files are tracked, so that when a xorb is uploaded, the progress of all the files containing data in that xorb is incremented correctly.  And, when a file is deduped against any xorb, that progress is also reported correctly.  

Updates are passed to a struct implementing the TrackingProgressUpdater trait, which reports progress as batches of 
per-file updates. 

```
pub trait TrackingProgressUpdater: Debug + Send + Sync {
    /// Register a set of updates as a list of ProgressUpdate instances, which 
    /// contain the name and progress information.
    fn register_updates(&self, updates: &[ProgressUpdate]);
}
```

where ProgressUpdate is defined as

```
pub struct ProgressUpdate {
    pub item_name: Arc<str>,
    pub total_count: u64,
    pub completed_count: u64,
    pub update_increment: u64,
}
```

In addition to the above functionality, numerous test cases were added to ensure accuracy and correctness across a lot of different cases.

assafvayner commented:
Need to resolve TODO's and ensure backwards compatibility while hf_hub keeps providing only 1 progress bar, then make a change to hf_hub to add progress bars for each file.

hoytak commented:
Updated the PR to the current main branch, so everything should work now. 

On further thought, I changed the trait for the progress updating to be async to avoid the use of the std::sync::Mutex; this should now be WASM compatible. 

The current implementation should be backwards compatible with the previous progress updating as reported to python.   

Let's do this in three stages:

1. Get this PR in.   This is backwards compatible; all the per-file updates are folded into the global updating counter. 
2. Update hub to have detailed progress reporting functions that work with the new file API.
3. A final PR here to connect the two.

assafvayner commented on merklehash/Cargo.toml:
```diff
@@ -6,7 +6,7 @@ edition = "2021"
 
 [dependencies]
-rand = "0.9.0"
+rand = { version = "0.8.5", features = ["small_rng"] }
```
upgrade rand version

assafvayner commented on hf_xet/src/progress_update.rs:
```diff
@@ -49,18 +49,27 @@ impl WrappedProgressUpdater {
-                .call1((increment,))
-                .log_error("python exception trying to update progress bar");
+            for update in updates {
+                // TODO: Connect the full information in the ProgressUpdate struct to python.
```
this is a TODO in the huggingface_hub library, we should remove this comment before merging or remove this comment after merging the PR into huggingface_hub. we have to release hf_xet before the huggingface_hub PR gets merged

hoytak replied:
The huggingface_hub would just expose a python API that should be called something different than this one, so that can get released before the connecting PR.   That connecting PR can clean up all these comments.

assafvayner reviewed (approved):
I think this is all good, would urge to get a DX or YL stamp too!

seanses commented on data/src/progress_tracking.rs:
```diff
@@ -0,0 +1,497 @@
+/// Represents a file that depends on one or more xorbs.
+struct FileDependency {
+    /// Human-readable name of the file.
+    name: Arc<str>,
```
Maybe we don't need this field. There's a trend for xet-core to support upload from any `Read` data source, e.g. https://github.com/huggingface/xet-core/pull/273/files

hoytak replied:
Since the progress bar functions as a callback, there needs to be some way of tracking which update is for which file.  This can just be a unique tag that the python end simply translates into the filename.

seanses replied:
A tag sounds good!

seanses commented on data/src/file_cleaner.rs:
```diff
@@ -39,7 +40,10 @@ enum DedupManagerBackgrounder {
 pub struct SingleFileCleaner {
     // Auxiliary info
-    file_name: String,
+    file_name: Arc<str>,
```
We can delete this field.

hoytak replied:
Again, the main issue here is that the callback gives per-file updates, and this field is needed to specify which updated applies to which file.  This can be a tag instead of a name that the end point translates, but needs to be something propagated through the whole process.

seanses commented:
Looks good in general, will take another look when the conflicts are resolved!
